File: apps/mobile/ios/CodexTracker/ConvexWireModels.swift
@preconcurrency import ConvexMobile
import Foundation

/// Convex encodes exceptional Float64 values as tagged objects. Keep that transport
/// representation out of fixture/domain models and normalize non-finite data at the boundary.
/// https://docs.convex.dev/client/swift/data-types#numerical-types
private func normalizedConvexNumber(_ value: Double) -> Double {
  value.isFinite ? value : 0
}

private func normalizedConvexNumber(_ value: Double?) -> Double? {
  value.map(normalizedConvexNumber)
}

struct ConvexCompactModelUsage: Decodable {
  let model: String
  let agent: String?
  @ConvexFloat var i: Double
  @ConvexFloat var c: Double
  @ConvexFloat var w: Double
  @ConvexFloat var o: Double
  @ConvexFloat var r: Double
  @ConvexFloat var t: Double
  @ConvexFloat var q: Double
  @ConvexFloat var usd: Double

  var domain: CompactModelUsage {
    CompactModelUsage(
      model: model,
      agent: agent,
      i: normalizedConvexNumber(i),
      c: normalizedConvexNumber(c),
      w: normalizedConvexNumber(w),
      o: normalizedConvexNumber(o),
      r: normalizedConvexNumber(r),
      t: normalizedConvexNumber(t),
      q: normalizedConvexNumber(q),
      usd: normalizedConvexNumber(usd)
    )
  }
}

struct ConvexCompactHourRow: Decodable {
  @ConvexFloat var h: Double
  let u: String
  let d: String
  @ConvexFloat var i: Double
  @ConvexFloat var c: Double
  @ConvexFloat var w: Double
  @ConvexFloat var o: Double
  @ConvexFloat var r: Double
  @ConvexFloat var t: Double
  @ConvexFloat var q: Double
  @ConvexFloat var usd: Double
  let m: [ConvexCompactModelUsage]

  var domain: CompactHourRow {
    CompactHourRow(
      h: normalizedConvexNumber(h),
      u: u,
      d: d,
      i: normalizedConvexNumber(i),
      c: normalizedConvexNumber(c),
      w: normalizedConvexNumber(w),
      o: normalizedConvexNumber(o),
      r: normalizedConvexNumber(r),
      t: normalizedConvexNumber(t),
      q: normalizedConvexNumber(q),
      usd: normalizedConvexNumber(usd),
      m: m.map(\.domain)
    )
  }
}

struct ConvexHourlyResponse: Decodable {
  let rows: [ConvexCompactHourRow]
  let users: [PublicUser]
}

struct ConvexLiveSnapshot: Decodable {
  let sessionId: String?
  let model: String?
  @ConvexFloat var tokensPerSecond: Double
  @OptionalConvexFloat var lastEventAt: Double?
  @ConvexFloat var todayTotal: Double
  @ConvexFloat var todayCost: Double
  @ConvexFloat var updatedAt: Double

  var domain: LiveSnapshot {
    LiveSnapshot(
      sessionId: sessionId,
      model: model,
      tokensPerSecond: normalizedConvexNumber(tokensPerSecond),
      lastEventAt: normalizedConvexNumber(lastEventAt),
      todayTotal: normalizedConvexNumber(todayTotal),
      todayCost: normalizedConvexNumber(todayCost),
      updatedAt: normalizedConvexNumber(updatedAt)
    )
  }
}

struct ConvexSessionItem: Decodable {
  let id: String
  let user: PublicUser
  let deviceId: String
  let sessionId: String
  let agent: String
  let model: String
  let projectName: String?
  @ConvexFloat var startedAt: Double
  @ConvexFloat var lastActivityAt: Double
  @ConvexFloat var input: Double
  @ConvexFloat var cached: Double
  @ConvexFloat var cacheWrite: Double
  @ConvexFloat var output: Double
  @ConvexFloat var reasoning: Double
  @ConvexFloat var total: Double
  @ConvexFloat var requests: Double
  @ConvexFloat var cost: Double
  let source: String?
  let cliVersion: String?

  var domain: SessionItem {
    SessionItem(
      id: id,
      user: user,
      deviceId: deviceId,
      sessionId: sessionId,
      agent: agent,
      model: model,
      projectName: projectName,
      startedAt: normalizedConvexNumber(startedAt),
      lastActivityAt: normalizedConvexNumber(lastActivityAt),
      input: normalizedConvexNumber(input),
      cached: normalizedConvexNumber(cached),
      cacheWrite: normalizedConvexNumber(cacheWrite),
      output: normalizedConvexNumber(output),
      reasoning: normalizedConvexNumber(reasoning),
      total: normalizedConvexNumber(total),
      requests: normalizedConvexNumber(requests),
      cost: normalizedConvexNumber(cost),
      source: source,
      cliVersion: cliVersion
    )
  }
}

struct ConvexMemberItem: Decodable {
  let id: String
  let name: String?
  let email: String?
  let imageUrl: String?
  let role: String
  @ConvexFloat var joinedAt: Double
  @ConvexFloat var deviceCount: Double
  @OptionalConvexFloat var lastSeenAt: Double?
  let live: ConvexLiveSnapshot?

  var domain: MemberItem {
    MemberItem(
      id: id,
      name: name,
      email: email,
      imageUrl: imageUrl,
      role: role,
      joinedAt: normalizedConvexNumber(joinedAt),
      deviceCount: normalizedConvexNumber(deviceCount),
      lastSeenAt: normalizedConvexNumber(lastSeenAt),
      live: live?.domain
    )
  }
}

struct ConvexDeviceItem: Decodable {
  let id: String
  let name: String
  let platform: String
  let hostname: String?
  let appVersion: String?
  let timezone: String?
  @ConvexFloat var createdAt: Double
  @ConvexFloat var lastSeenAt: Double
  let live: ConvexLiveSnapshot?
  @ConvexFloat var logins: Double

  var domain: DeviceItem {
    DeviceItem(
      id: id,
      name: name,
      platform: platform,
      hostname: hostname,
      appVersion: appVersion,
      timezone: timezone,
      createdAt: normalizedConvexNumber(createdAt),
      lastSeenAt: normalizedConvexNumber(lastSeenAt),
      live: live?.domain,
      logins: normalizedConvexNumber(logins)
    )
  }
}

struct ConvexLiveItem: Decodable {
  let user: PublicUser
  let deviceId: String
  let deviceName: String
  let platform: String
  let live: ConvexLiveSnapshot

  var domain: LiveItem {
    LiveItem(
      user: user,
      deviceId: deviceId,
      deviceName: deviceName,
      platform: platform,
      live: live.domain
    )
  }
}

File: apps/mobile/android/app/src/main/java/dev/chenli/codextracker/data/ConvexWireModels.kt
package dev.chenli.codextracker.data

import dev.chenli.codextracker.domain.Account
import dev.chenli.codextracker.domain.CompactHourRow
import dev.chenli.codextracker.domain.CompactModelUsage
import dev.chenli.codextracker.domain.Device
import dev.chenli.codextracker.domain.HourlyResponse
import dev.chenli.codextracker.domain.LiveDevice
import dev.chenli.codextracker.domain.LiveSnapshot
import dev.chenli.codextracker.domain.Member
import dev.chenli.codextracker.domain.Organization
import dev.chenli.codextracker.domain.PublicUser
import dev.chenli.codextracker.domain.QueryRange
import dev.chenli.codextracker.domain.UsageScope
import dev.chenli.codextracker.domain.UsageSession
import dev.convex.android.Float64
import kotlinx.serialization.Serializable

private const val MaxSafeInteger = 9_007_199_254_740_991.0

internal fun Double.toDomainLong(field: String): Long {
  require(isFinite()) { "$field must be finite" }
  require(this >= 0.0) { "$field must be non-negative" }
  require(this <= MaxSafeInteger) { "$field exceeds the exact JavaScript integer range" }
  require(this % 1.0 == 0.0) { "$field must be a whole number" }
  return toLong()
}

internal fun Double.toDomainInt(field: String): Int {
  val value = toDomainLong(field)
  require(value <= Int.MAX_VALUE) { "$field exceeds the Int range" }
  return value.toInt()
}

private fun Double.toDomainMeasure(field: String): Double {
  require(isFinite()) { "$field must be finite" }
  require(this >= 0.0) { "$field must be non-negative" }
  return this
}

private fun Long.toConvexNumber(field: String): Double {
  require(toDouble() in -MaxSafeInteger..MaxSafeInteger) {
    "$field exceeds the exact JavaScript integer range"
  }
  return toDouble()
}

internal object LiveQueryArguments {
  fun hourly(
    scope: UsageScope,
    orgId: String?,
    range: QueryRange,
  ): Map<String, Any?> =
    scoped(scope, orgId) +
      mapOf(
        "from" to range.from.toConvexNumber("from"),
        "to" to range.to.toConvexNumber("to"),
      )

  fun recentSessions(
    scope: UsageScope,
    orgId: String?,
    limit: Int,
  ): Map<String, Any?> {
    require(limit in 1..100) { "limit must be between 1 and 100" }
    return scoped(scope, orgId) + mapOf("limit" to limit.toDouble())
  }

  fun scoped(scope: UsageScope, orgId: String?): Map<String, Any?> =
    buildMap {
      put("scope", scope.name.lowercase())
      if (scope == UsageScope.Team && orgId != null) put("orgId", orgId)
    }
}

@Serializable
internal data class ConvexAccount(
  val id: String,
  val name: String? = null,
  val email: String? = null,
  val imageUrl: String? = null,
  val onboardedAt: Float64? = null,
) {
  fun toDomain() =
    Account(
      id = id,
      name = name,
      email = email,
      imageUrl = imageUrl,
      onboardedAt = onboardedAt?.toDomainLong("account.onboardedAt"),
    )
}

@Serializable
internal data class ConvexCompactModelUsage(
  val model: String,
  val agent: String? = null,
  val i: Float64,
  val c: Float64,
  val w: Float64,
  val o: Float64,
  val r: Float64,
  val t: Float64,
  val q: Float64,
  val usd: Float64,
) {
  fun toDomain() =
    CompactModelUsage(
      model = model,
      agent = agent,
      i = i.toDomainLong("model.i"),
      c = c.toDomainLong("model.c"),
      w = w.toDomainLong("model.w"),
      o = o.toDomainLong("model.o"),
      r = r.toDomainLong("model.r"),
      t = t.toDomainLong("model.t"),
      q = q.toDomainLong("model.q"),
      usd = usd.toDomainMeasure("model.usd"),
    )
}

@Serializable
internal data class ConvexCompactHourRow(
  val h: Float64,
  val u: String,
  val d: String,
  val i: Float64,
  val c: Float64,
  val w: Float64,
  val o: Float64,
  val r: Float64,
  val t: Float64,
  val q: Float64,
  val usd: Float64,
  val m: List<ConvexCompactModelUsage>,
) {
  fun toDomain() =
    CompactHourRow(
      h = h.toDomainLong("hour.h"),
      u = u,
      d = d,
      i = i.toDomainLong("hour.i"),
      c = c.toDomainLong("hour.c"),
      w = w.toDomainLong("hour.w"),
      o = o.toDomainLong("hour.o"),
      r = r.toDomainLong("hour.r"),
      t = t.toDomainLong("hour.t"),
      q = q.toDomainLong("hour.q"),
      usd = usd.toDomainMeasure("hour.usd"),
      m = m.map(ConvexCompactModelUsage::toDomain),
    )
}

@Serializable
internal data class ConvexHourlyResponse(
  val rows: List<ConvexCompactHourRow>,
  val users: List<PublicUser>,
) {
  fun toDomain() = HourlyResponse(rows.map(ConvexCompactHourRow::toDomain), users)
}

@Serializable
internal data class ConvexLiveSnapshot(
  val sessionId: String? = null,
  val model: String? = null,
  val tokensPerSecond: Float64,
  val lastEventAt: Float64? = null,
  val todayTotal: Float64,
  val todayCost: Float64,
  val updatedAt: Float64,
) {
  fun toDomain() =
    LiveSnapshot(
      sessionId = sessionId,
      model = model,
      tokensPerSecond = tokensPerSecond.toDomainMeasure("live.tokensPerSecond"),
      lastEventAt = lastEventAt?.toDomainLong("live.lastEventAt"),
      todayTotal = todayTotal.toDomainLong("live.todayTotal"),
      todayCost = todayCost.toDomainMeasure("live.todayCost"),
      updatedAt = updatedAt.toDomainLong("live.updatedAt"),
    )
}

@Serializable
internal data class ConvexUsageSession(
  val id: String,
  val user: PublicUser,
  val deviceId: String,
  val sessionId: String,
  val agent: String,
  val model: String,
  val projectName: String? = null,
  val startedAt: Float64,
  val lastActivityAt: Float64,
  val input: Float64,
  val cached: Float64,
  val cacheWrite: Float64,
  val output: Float64,
  val reasoning: Float64,
  val total: Float64,
  val requests: Float64,
  val cost: Float64,
  val source: String? = null,
  val cliVersion: String? = null,
) {
  fun toDomain() =
    UsageSession(
      id = id,
      user = user,
      deviceId = deviceId,
      sessionId = sessionId,
      agent = agent,
      model = model,
      projectName = projectName,
      startedAt = startedAt.toDomainLong("session.startedAt"),
      lastActivityAt = lastActivityAt.toDomainLong("session.lastActivityAt"),
      input = input.toDomainLong("session.input"),
      cached = cached.toDomainLong("session.cached"),
      cacheWrite = cacheWrite.toDomainLong("session.cacheWrite"),
      output = output.toDomainLong("session.output"),
      reasoning = reasoning.toDomainLong("session.reasoning"),
      total = total.toDomainLong("session.total"),
      requests = requests.toDomainLong("session.requests"),
      cost = cost.toDomainMeasure("session.cost"),
      source = source,
      cliVersion = cliVersion,
    )
}

@Serializable
internal data class ConvexMember(
  val id: String,
  val name: String? = null,
  val email: String? = null,
  val imageUrl: String? = null,
  val role: String,
  val joinedAt: Float64,
  val deviceCount: Float64,
  val lastSeenAt: Float64? = null,
  val live: ConvexLiveSnapshot? = null,
) {
  fun toDomain() =
    Member(
      id = id,
      name = name,
      email = email,
      imageUrl = imageUrl,
      role = role,
      joinedAt = joinedAt.toDomainLong("member.joinedAt"),
      deviceCount = deviceCount.toDomainInt("member.deviceCount"),
      lastSeenAt = lastSeenAt?.toDomainLong("member.lastSeenAt"),
      live = live?.toDomain(),
    )
}

@Serializable
internal data class ConvexDevice(
  val id: String,
  val name: String,
  val platform: String,
  val hostname: String? = null,
  val appVersion: String? = null,
  val timezone: String? = null,
  val createdAt: Float64,
  val lastSeenAt: Float64,
  val live: ConvexLiveSnapshot? = null,
  val logins: Float64,
) {
  fun toDomain() =
    Device(
      id = id,
      name = name,
      platform = platform,
      hostname = hostname,
      appVersion = appVersion,
      timezone = timezone,
      createdAt = createdAt.toDomainLong("device.createdAt"),
      lastSeenAt = lastSeenAt.toDomainLong("device.lastSeenAt"),
      live = live?.toDomain(),
      logins = logins.toDomainInt("device.logins"),
    )
}

@Serializable
internal data class ConvexLiveDevice(
  val user: PublicUser,
  val deviceId: String,
  val deviceName: String,
  val platform: String,
  val live: ConvexLiveSnapshot,
) {
  fun toDomain() =
    LiveDevice(
      user = user,
      deviceId = deviceId,
      deviceName = deviceName,
      platform = platform,
      live = live.toDomain(),
    )
}

@Serializable
internal data class ConvexBackendOrganization(
  val id: String,
  val name: String,
  val slug: String? = null,
  val imageUrl: String? = null,
  val role: String,
  val memberCount: Float64,
) {
  fun toDomain(clerkOrgId: String): Organization {
    memberCount.toDomainInt("organization.memberCount")
    return Organization(
      id = id,
      clerkOrgId = clerkOrgId,
      name = name,
      slug = slug,
      imageUrl = imageUrl,
      role = role,
    )
  }
}
